### ISSUE TITLE
Possible memory issue on the master nodes in 5.5.1 with the large amount of idexes/shards

### ISSUE THREAD

**Issue by lruslan** [elastic/elasticsearch#26610]
**Elasticsearch version** (`bin/elasticsearch --version`): 5.5.1, build[19c13d0/2017-07-18T20:44:24.823Z]

**Plugins installed**: [prometheus-exporter]

**JVM version** (`java -version`): OpenJDK Runtime Environment (build 1.8.0_131-8u131-b11-2ubuntu1.16.04.3-b11)

**OS version** (`uname -a` if on a Unix-like system): Linux 4.4.0-78-generic #99-Ubuntu SMP Thu Apr 27 15:29:09 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

**Description of the problem including expected versus actual behavior**:

Memory usage keeps increasing on the master nodes. We see steady growth 5Gb per 24 hours. Such behaviour started as we increased number of data nodes and switched from weekly to hourly index rotation - in old setup with 5 data nodes and daily index rotation size of shards was relatively small <100 and it was enough to have master with 8Gb of RAM, now with 20 data nodes and > 400indexes with 20 shards per index we have >8,000 shards and it seem even 14Gb of RAM on the master is not enough. It's just keeps growing until OOM. In our setup we use separate dedicated nodes: masters (3 nodes) , data nodes (20 nodes), coordinator nodes (4 nodes) It's only master nodes have such issue, memory usage on other types of nodes are stable and properly garbage collected.
 
Java Heap Stats:
<img width="1249" alt="screen shot 2017-09-12 at 21 10 54" src="https://user-images.githubusercontent.com/1275141/30344674-8ef506a6-9802-11e7-9a22-2ed7b09a8cde.png">

Top classes from jhat processed heap dump of the master (with 8GB of RAM used)
Sorted by size:
<img width="1087" alt="screen shot 2017-09-12 at 21 03 16" src="https://user-images.githubusercontent.com/1275141/30344794-053647a8-9803-11e7-9a14-b556d18c0d48.png">
Sorted by count:
<img width="1105" alt="screen shot 2017-09-12 at 21 04 19" src="https://user-images.githubusercontent.com/1275141/30344824-25f1b4f0-9803-11e7-90ca-daee00ff205d.png">

Configuration of the master:
```
cluster.name: logging5
node.name: l5-es-m-1h5
node.master: true
node.data: false
node.ingest: false
network.host: 172.6.8.24
discovery.zen.minimum_master_nodes: 2
discovery.zen.ping.unicast.hosts: ["172.16.8.23","172.16.8.24","172.16.8.25"]
action.destructive_requires_name: true
```

**Provide logs (if relevant)**:
In the logs on the master I see constant updates of certain dynamic mappings, but we always had similar updates and they never caused memory related issues, especially on master nodes:
```
[2017-09-12T21:34:35,672][INFO ][o.e.c.m.MetaDataMappingService] [l5-es-m-0i9] [logstash-2017.09.12.19/3J9B1jZXQROU1FOTbWiFBQ] update_mapping [v2]
[2017-09-12T21:36:24,578][INFO ][o.e.c.m.MetaDataMappingService] [l5-es-m-0i9] [logstash-2017.09.12.19/3J9B1jZXQROU1FOTbWiFBQ] update_mapping [v2]
[2017-09-12T21:39:05,945][INFO ][o.e.c.m.MetaDataMappingService] [l5-es-m-0i9] [logstash-2017.09.12.19/3J9B1jZXQROU1FOTbWiFBQ] update_mapping [v2]
[2017-09-12T21:39:07,829][INFO ][o.e.c.m.MetaDataMappingService] [l5-es-m-0i9] [logstash-2017.09.12.19/3J9B1jZXQROU1FOTbWiFBQ] update_mapping [v2]
[2017-09-12T21:40:05,012][INFO ][o.e.c.m.MetaDataMappingService] [l5-es-m-0i9] [logstash-2017.09.12.19/3J9B1jZXQROU1FOTbWiFBQ] update_mapping [v2]
[2017-09-12T21:41:34,597][INFO ][o.e.c.m.MetaDataMappingService] [l5-es-m-0i9] [logstash-2017.09.12.19/3J9B1jZXQROU1FOTbWiFBQ] update_mapping [v2]
[2017-09-12T21:43:04,053][INFO ][o.e.c.m.MetaDataMappingService] [l5-es-m-0i9] [logstash-2017.09.12.19/3J9B1jZXQROU1FOTbWiFBQ] update_mapping [k8s] 
```

In case you need it I can provide access to the complete heap dump.

**Comment by ywelsch** [comment]
- Is this something you're newly experiencing with 5.5.1 but worked fine in earlier ES versions?
- how many total shards you have?
- what's the size of your cluster state?
- are you experiencing the same issues when the `prometheus-exporter` plugin is not installed?
- Can you provide a heap dump that's taken on OOM?

**Comment by colings86** [comment]
8,000 shards across 20 data nodes is quite excessive, you should consider reducing the number of shards per index as it is likely you are adversely affecting the performance of your cluster with this. Also what was the reason for moving form weekly to hourly indices? What is the average size on disk for your shards with hourly indices? Each shard has an overhead both on the node that it is allocated to and on the master node (in the form of entries in the cluster state). With the kinds of numbers you are talking about here you may well be getting memory problems due to the shear size of you cluster state, especially if you indices have a reasonably large number of fields. I would consider going back to daily or weekly indices here.

**Comment by lruslan** [comment]
@ywelsch 

- Is this something you're newly experiencing with 5.5.1 but worked fine in earlier ES versions?
No. Before we were mostly using 1.7.* with considerably smaller amount of data nodes/indexes. And it's only in 5.5.1 we changed cluster setup to use 20 datanodes  and hourly index.
- how many total shards you have?
9087 shards ( 31 nodes in cluster, 20 data nodes, 490 indexes, 20 shards per index, hourly index) 
- what's the size of your cluster state?
64MB
- are you experiencing the same issues when the prometheus-exporter plugin is not installed?
yes we experienced same memory growth issues before the  prometheus-exporter plugin, I actually installed prometheus-exporter on master nodes to start collecting jvm stats
- Can you provide a heap dump that's taken on OOM?
For now I can provide dump from the jvm with memory usage close to OOM (8G of heap used out of 12G)
https://drive.google.com/a/travix.com/file/d/0B40JaBxFEd3ET3FMRDhmY0FzbkE/view?usp=sharing

 I will try to get another taken during OOM will let you know when I have it.

Thank you

**Comment by lruslan** [comment]
@colings86 thanks for the suggestions,
- it is likely you are adversely affecting the performance of your cluster with this
With current setup our performance is actually very good (I tried number of different setups) it's only masters expose this memory issue (btw CPU usage on all the servers is really low). I think we may even go up to 30G of RAM on masters, maybe memory usage will settle somewhere close to 20G  (would be useful to know if there is a formula to estimate how much memory would be needed by masters assuming we have this many shards with this many fields). My initial assumption was - masters are not doing any particular heavy lifting and should be fine with relatively low resources assigned.  Initially I tried with the smaller amount of data nodes/shards which resulted in slower searches larger timeouts  and our systems being under utilised - we are running on dedicated hardware servers with 256G of RAM and 48 cores. Now I have 4 instances of elasticsearch per hardware server running in docker and with 20 shards I see huge improvement on a search/indexing side.
- what was the reason for moving form weekly to hourly indices?
we write almost 1.2-1.4TB of data daily to our logstash indexes, with daily index and 20 data nodes we will have approximately 60GB-80Gb per shard (first we tried daily index but it proved too slow to recover/relocate quickly in case of hardware issues, curator index merging on 60G takes too long, also with daily index we are not so flexible when need adjust mappings/change type of filed during the day - we need to edit index template and wait another day/index to start using changed template, with hourly index it's way too faster)
So far it seems hourly index is our best bet. What I wanted to do next is reduce amount of shards from 20 to 10 but then enable replication (to have at least one replica) which will bring us to the same amount of shards 20 ( 10 index + 10 replica)
- What is the average size on disk for your shards with hourly indices?
3.7Gb

**Comment by ywelsch** [comment]
> For now I can provide dump from the jvm with memory usage close to OOM (8G of heap used out of 12G)
https://drive.google.com/a/travix.com/file/d/0B40JaBxFEd3ET3FMRDhmY0FzbkE/view?usp=sharing

I can't access the heap dump. Can you give permission to `yannick _AT_ elastic.co`?

**Comment by ywelsch** [comment]
The heapdump contains mostly garbage (waiting to be collected), 400MB out of 6,7GB are actually reachable objects. Only a heapdump on OOM can explain what's going on here.

**Comment by lruslan** [comment]
@ywelsch thanks. I've enabled HeapDumpOnOutOfMemoryError and gc logging , will let you know once I have a heapdump on OOM. 
I also can see heap garbage being properly collected once in a while with 10G heap limit, jvm allows heap grow up to  7G and then collection happens. It looks I rushed to post an issue which is not real problem.
<img width="1239" alt="screen shot 2017-09-13 at 15 59 39" src="https://user-images.githubusercontent.com/1275141/30381744-92cd487e-989d-11e7-87fb-a70e6911dd89.png">
